languages/csharp/exercises/concept/user-defined-exceptions/UserDefinedExceptons.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+
+public class CalculationException_template : Exception
+{
+    private int operand;
+
+    public CalculationException_template(int operand, string message, Exception inner)
+    // TODO: complete the definition of the constructor
+    {
+
+    }
+
+    public int GetOperand()
+    {
+        throw new NotImplementedException($"Please implement the GetOperand() method");
+    }
+}
+
+public class CalculatorTestHarness_template
+{
+    public string Run(string testName, int testValue)
+    {
+        throw new NotImplementedException($"Please implement the Run() method");
+    }
+
+    public void Calculate(int testNum)
+    {
+    }
+}
+
+// Please do not modify the code below.  In a more realistic
+// scenario this would be in a different library and would actually
+// do something meaningful.  This code will always throw
+// a Sysem.OverflowException
+public static class TestHarnessOperations_template
+{
+    public static int HandleInt(int testNum)
+    {
+        int trouble = Int32.MaxValue;
+        checked
+        {
+            return Int32.MaxValue * trouble;
+        }
+    }
+}
```

Comment:
I had some trouble understanding what exactly I was expected to write as a student. What if we would do the following, where the test harness wraps a calculator class:

```csharp
public class CalculatorTestHarness
{
    private readonly Calculator calculator;

    public CalculatorTestHarness(Calculator calculator)
    {
        this.calculator = calculator;
    }
    
    public string TestMultiplication(int x, int y)
    {
        throw new NotImplementedException($"Please implement the TestMultiplication() method");
    }
}

public class Calculator
{
    public int Multiplication(int x, int y)
    {
        checked
        {
            return x * y;
        }
    }
}
```

We could then provide the multiplication values ourselves to trigger the overflow. What do you think?

Reply:
1. I avoided the possibility that the lowest level routine would fail to fail as I was concerned that the user might find a way of guarding against it in the calling routine but on balance this looks better.
2. We will have to add the `const` exercise as a pre-requisite.  I had it in mind to minimise dependencies (hence avoiding the use of a property as well) to limit administrative hassles apart from anything else as we don't have a tree yet. 
3. I assume we still want a routine to wrap and throw and a different routine to catch and filter.  I think a certain amount of confusion comes in having two routines doing something very similar so any ideas here would not go amiss.

Reply:
> 1. I avoided the possibility that the lowest level routine would fail to fail as I was concerned that the user might find a way of guarding against it in the calling routine but on balance this looks better.

Good to hear!

> 2. We will have to add the const exercise as a pre-requisite. I had it in mind to minimise dependencies (hence avoiding the use of a property as well) to limit administrative hassles apart from anything else as we don't have a tree yet and where the exercise has not been written.

Ah, let's remove the `readonly` bit to reduce dependencies. Good call. I think the properties dependency can help the student a bit here to remove some of the boilerplate, but I don't mind keeping the method instead of the property. You can convince me either way.

> 3. I assume we still want a routine to wrap and throw and a different routine to catch and filter. I think a certain amount of confusion comes in having two routines doing something very similar so any ideas here would not go amiss.

This might indeed be a good idea. So adding a `Multiplication` method to the `CalculatorTestHarness` class:

```csharp
public int Multiplication(int x, int y)
{
    ...
}
```